assembly/src/assembler/instruction/procedures.rs

```diff
@@ -78,51 +81,86 @@ impl Assembler {
             None => context.register_phantom_call(Span::new(span, mast_root))?,
         }
 
-        let block = match kind {
-            // For `exec`, we use a PROXY block to reflect that the root is
-            // conceptually inlined at this location
-            InvokeKind::Exec => CodeBlock::new_proxy(mast_root),
-            // For `call`, we just use the corresponding CALL block
-            InvokeKind::Call => CodeBlock::new_call(mast_root),
-            // For `syscall`, we just use the corresponding SYSCALL block
-            InvokeKind::SysCall => CodeBlock::new_syscall(mast_root),
+        let mast_root_node_id = {
+            // Note that here we rely on the fact that we topologically sorted the procedures, such
+            // that when we assemble a procedure, all procedures that it calls will have been
+            // assembled, and hence be present in the `MastForest`. We currently assume that the
+            // `MastForest` contains all the procedures being called; "external procedures" only
+            // known by digest are not currently supported.
+            let callee_id = mast_forest
```

Comment:
I may be misremembering, but I thought the conclusion we landed on in the call was to check if the digest is in the `MastForest`, and if so, whether it is a `Proxy`. If it is not in the `MastForest`, or resolves to a `Proxy` node, then the callee must be external, and we can safely use a `Proxy` node without concern for conflicting with whatever node type the proxy refers to. As mentioned in your comment, we can rely on the fact that all callees are processed before callers to guarantee that proxies will never conflict with "real" nodes (since they represent mutually exclusive states).

WIth the above in mind, rather than raising an error on proxy nodes (which would be a major issue for `call` in particular, as it will almost always be the case that the callee for `call` is _not_ in the current `MastForest`), we would create them only when the proxied node isn't available in the current forest. Previously, we were just always using `Proxy`, even if the proxied node might be in the current `MastForest`, so that's the main change IIRC.

This sidesteps the issue of `Proxy` nodes having the same hash as the proxied node (since they can't exist simultaneously, the fact that they conflict is moot), and vastly (or perhaps even entirely, in some cases) reduces the number of proxy nodes emitted.

Let me know if I've gotten that wrong, I would like to refresh my memory on the plan of attack here if I've misremembered things.